src/libstore/misc.cc

```diff
@@ -169,11 +169,11 @@ Paths Store::topoSortPaths(const PathSet & paths)
     Paths sorted;
     PathSet visited, parents;
 
-    std::function<void(const Path & path)> dfsVisit;
+    std::function<void(const Path & path, const Path * parent)> dfsVisit;
 
-    dfsVisit = [&](const Path & path) {
+    dfsVisit = [&](const Path & path, const Path * parent) {
         if (parents.find(path) != parents.end())
-            throw BuildError(format("cycle detected in the references of ‘%1%’") % path);
+            throw BuildError(format("cycle detected in the references of ‘%1%’ from '%2%'") % path % *parent);
```

Comment:
@edolstra I use a pointer here as `Maybe` (and hope there's a better solution, I just don't know one...). Either way this code should be safe:

1. On any first iteration of DFS (when `parent` is `nullptr`) `parents` is empty:
  a. by construction on the first call;
  b. because the only place when `parents` is modified is in the beginning of `dfsVisit` when the path is inserted, and in the end when deleted: `dfsVisit` leaves `parents` _in the same state_ it started with, so after any top-level call of `dfsVisit` `parents` is empty again.
2. On all the following iterations `parent` is non-null, so it's safe.